compiler/rustc_ast/src/ast.rs

```diff
@@ -3316,6 +3316,42 @@ impl VisibilityKind {
     }
 }
 
+#[derive(Clone, Encodable, Decodable, Debug)]
+pub struct Restriction {
+    pub kind: RestrictionKind,
+    pub span: Span,
+    pub tokens: Option<LazyAttrTokenStream>,
+}
+
+#[derive(Clone, Encodable, Decodable, Debug)]
+pub enum RestrictionKind {
+    Unrestricted,
+    Restricted { path: P<Path>, id: NodeId, shorthand: bool },
+    Implied,
+}
+
+impl Restriction {
+    pub fn unrestricted() -> Self {
+        Restriction { kind: RestrictionKind::Unrestricted, span: DUMMY_SP, tokens: None }
+    }
+
+    pub fn restricted(path: P<Path>, id: NodeId, shorthand: bool) -> Self {
+        Restriction {
+            kind: RestrictionKind::Restricted { path, id, shorthand },
+            span: DUMMY_SP,
+            tokens: None,
+        }
+    }
+
+    pub fn implied() -> Self {
+        Restriction { kind: RestrictionKind::Implied, span: DUMMY_SP, tokens: None }
+    }
```

Comment:
Looking at the previous PR, https://github.com/rust-lang/rust/pull/106074/files#diff-68dedde73c3f169dfeeed73fc362bede2bb414477a495b8b2fb626fc5e91dd19R172 seems to be the only place where a `Span` is not given, and it's for `mut` restriction, let's deal with that later.

Could you change the methods so take `Span` directly, that should avoid the use of `DUMMY_SP` and should avoid accidental omition of the span.